CDAP-7683 add a tool to upload coprocessor jars

albertshau opened the pull request:
Added a tool that will build and upload coprocessor jars to HDFS.
The tool will run before the master starts, and can also be
run on its own. It can be used on a slave cluster to ensure that
required coprocessor jars are on HDFS even though CDAP is not
running on the cluster.

Refactored the logic around building coprocessors so that
HBaseTableAdmin and HBaseQueueAdmin no longer build their required
coprocessor on demand, but just assume the coprocessor exists at
a pre-defined location. Logic for building the actual jar was
just moved from HBaseTableUtil to a new CoprocessorManager class.

albertshau committed:
CDAP-7683 add a tool to upload coprocessor jars

Added a tool that will build and upload coprocessor jars to HDFS.
The tool will run before the master starts, and can also be
run on its own. It can be used on a slave cluster to ensure that
required coprocessor jars are on HDFS even though CDAP is not
running on the cluster.

Refactored the logic around building coprocessors so that
HBaseTableAdmin and HBaseQueueAdmin no longer build their required
coprocessor on demand, but just assume the coprocessor exists at
a pre-defined location. Logic for building the actual jar was
just moved from HBaseTableUtil to a new CoprocessorManager class.

albertshau commented:
http://builds.cask.co/browse/CDAP-DUT5318

sagarkapare commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/CoprocessorManager.java:
```diff
@@ -0,0 +1,196 @@
+ */
+public class CoprocessorManager {
+  private static final Logger LOG = LoggerFactory.getLogger(CoprocessorManager.class);
+  protected final Location jarDir;
```
nit: `private`

sagarkapare commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/CoprocessorManager.java:
```diff
@@ -0,0 +1,196 @@
+
+  /**
+   * Get the location of the specified type of coprocessor and ensure it exists.
+   * In distributed mode, the coprocessor jars are loaded onto hdfs by the CoprocessBuildTool,
```
nit: `CoprocessorBuildTool`

sagarkapare commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/CoprocessorManager.java:
```diff
@@ -0,0 +1,196 @@
+      return null;
+    }
+
+    LOG.debug("Creating jar file for coprocessor classes: " + buf.toString());
```
nit: use `{}` syntax with LOG.

sagarkapare commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/CoprocessorManager.java:
```diff
@@ -0,0 +1,196 @@
+
+    LOG.debug("Creating jar file for coprocessor classes: " + buf.toString());
+    final Hasher hasher = Hashing.md5().newHasher();
+    final byte[] buffer = new byte[4 * 1024];
```
better to move above two declarations close to where they are used. i.e. below the following for loop

albertshau replied:
the hasher isn't actually used anymore either, I'll remove it.

sagarkapare commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/CoprocessorManager.java:
```diff
@@ -0,0 +1,196 @@
+   * @param type the type of coprocessor
+   * @return the location of the coprocessor
+   * @throws IOException if there was an issue accessing the location
+   * @throws IllegalStateException if the coprocessor jar does not exist
```
where are we throwing `IllegalStateException`?

albertshau replied:
sorry, was originally there but removed it. will remove this line.

sagarkapare commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/CoprocessorManager.java:
```diff
@@ -0,0 +1,196 @@
+      buf.append(c.getName()).append(", ");
+    }
+    if (buf.length() == 0) {
+      return null;
```
method should be marked `@Nullable`

albertshau replied:
this should never happen anymore, will remove it.

sagarkapare commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/CoprocessorManager.java:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    // create the coprocessor jar on local filesystem
+    LOG.debug("Adding " + dependentClasses.size() + " classes to jar");
```
nit: use `{}` syntax

sagarkapare commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java:
```diff
@@ -33,13 +33,9 @@
-import com.google.common.base.Throwables;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Maps;
 import com.google.common.hash.Hasher;
```
unused import?

albertshau replied:
it was only used in the method to create the coprocessor jar, which got moved

sagarkapare reviewed (commented):
Minor nits. Rest LGTM 👍

dereklwood commented on cdap-common/bin/functions.sh:
```diff
@@ -643,6 +643,14 @@ cdap_start_java() {
+
+    # Build and upload coprocessor jars
+    logecho "$(date) Ensuring required HBase coprocessors are on HDFS"
+    "${JAVA}" ${JAVA_HEAPMAX} ${OPTS} -cp ${CLASSPATH} co.cask.cdap.data.tools.CoprocessorBuildTool </dev/null >>${__logfile} 2>&1
```
can this not just call the new ``cdap_setup_tool`` function to do the same thing?

albertshau replied:
the problem is it would re-run all the functions that set classpath and everything.

dereklwood replied:
Also note that if this is desirable in a CM environment, a corresponding update is needed to the CSD control.sh script (it does not use ``cdap_start_java``)

albertshau replied:
Is that script in this repo? Can't seem to find it.

dereklwood replied:
no, sorry, separate repo:  https://github.com/caskdata/cm_csd/blob/develop/src/scripts/cdap-control.sh

wolf31o2 replied:
Those functions *should* be no-op/idempotent, if the variables are already set. If they're not, it's a bug. It's better to setup the function in a way that we reuse it, versus having the same code in multiple places.

albertshau replied:
ok cool i'll try that then

albertshau replied:
i'll open a jira for the CM work

albertshau committed:
CDAP-7683 cleanup coprocessor creation code

wolf31o2 commented on cdap-common/bin/cdap:
```diff
@@ -74,6 +74,7 @@ case ${1} in
   upgrade) shift; cdap_upgrade_tool ${@}; __ret=${?} ;;
   run) shift; cdap_run_class ${@}; __ret=${?} ;;
   sdk) shift; cdap_sdk ${@}; __ret=${?} ;;
+  setup) shift; cdap_setup_tool ${@}; __ret=${?} ;;
```
This name is really ambiguous, especially given this is a very specific use case. Maybe `hbase_setup` or something, instead? This makes it seem like it's used for more than just adding coprocessors to HBase.

wolf31o2 commented on cdap-common/bin/cdap:
```diff
@@ -92,6 +93,7 @@ case ${1} in
     echo "    classpath    - Returns the Java CLASSPATH of the CDAP Master service"
     echo "    run          - Runs a given class with arguments using the CDAP Master CLASSPATH and environment"
     echo "    upgrade      - Runs the CDAP Upgrade with optional arguments"
+    echo "    setup        - Ensures CDAP coprocessors are present on HDFS"
```
HBase coprocessors

wolf31o2 commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/CoprocessorManager.java:
```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
```
2017

wolf31o2 commented on cdap-master/src/main/java/co/cask/cdap/data/tools/CoprocessorBuildTool.java:
```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
```
2017

wolf31o2 reviewed (changes requested):
Couple changes

albertshau committed:
CDAP-7683 script improvements

wolf31o2 commented on cdap-common/bin/cdap:
```diff
@@ -74,7 +74,7 @@ case ${1} in
   run) shift; cdap_run_class ${@}; __ret=${?} ;;
   sdk) shift; cdap_sdk ${@}; __ret=${?} ;;
-  setup) shift; cdap_setup_tool ${@}; __ret=${?} ;;
+  setup) shift; cdap_hbase_setup ${@}; __ret=${?} ;;
```
I was thinking of the argument being more `hbase_setup` than the function name. Sorry that I wasn't clear.

However, this isn't something that Nitin likes. You can see on the [wiki](https://wiki.cask.co/display/CE/CDAP+Scripts) that we're using compound commands, versus underscored ones. This was something that he specifically had me change in the other implementations, so I think we'll need to apply that here, too.

Maybe this should be `cdap setup hbase`?

```bash
  setup) shift; cdap_setup ${@}; __ret=${?} ;;
```
Then, the `cdap_setup` function would be something like:
```bash
cdap_setup() {
  local __action=${1}
  shift
  cdap_setup_${__action} ${@}
  return $?
}
```

albertshau replied:
ah ok got it

wolf31o2 commented on cdap-common/bin/functions.sh:
```diff
@@ -1088,11 +1084,12 @@ cdap_upgrade_tool() {
 
 #
-# cdap_setup_tool [arguments]
+# cdap_hbase_setup [arguments]
```
If using ideas from above: `cdap_setup_hbase`

wolf31o2 commented on cdap-common/bin/functions.sh:
```diff
@@ -1088,11 +1084,12 @@ cdap_upgrade_tool() {
+cdap_hbase_setup() {
   local readonly __ret __class=co.cask.cdap.data.tools.CoprocessorBuildTool
 
+  logecho "$(date) Ensuring required HBase coprocessors are on HDFS"
```
Move this up to where we're calling `cdap_hbase_setup` in `cdap_start_java` so we don't write to the log when called on the command line.

wolf31o2 reviewed (changes requested):
One kinda major change... this code is fine, as is... but I think we should make the change now to fit with what's been approved before.

albertshau committed:
CDAP-7683 change command to cdap setup coprocessors

Changing the script around so that setup takes a component and
arguments rather than being a single command. Also renaming
hbase to coprocessors to be more specific, and making sure the
tool output is written to the logfile when starting cdap instead
of printed to stdout.
